src/Catalyst.Core.Modules.Dfs/BlockExchange/BitSwapService.cs

```diff
@@ -0,0 +1,557 @@
+            }
+
+            foreach (var cid in _wants.Keys)
+            {
+                Unwant(cid);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///   The blocks needed by the peer.
+        /// </summary>
+        /// <param name="peer">
+        ///   The unique ID of the peer.
+        /// </param>
+        /// <returns>
+        ///   The sequence of CIDs need by the <paramref name="peer"/>.
+        /// </returns>
+        public IEnumerable<Cid> PeerWants(MultiHash peer)
+        {
+            return _wants.Values
+               .Where(w => w.Peers.Contains(peer))
+               .Select(w => w.Id);
+        }
+
+        /// <summary>
+        ///   Adds a block to the want list.
+        /// </summary>
+        /// <param name="id">
+        ///   The CID of the block to add to the want list.
+        /// </param>
+        /// <param name="peer">
+        ///   The unique ID of the peer that wants the block.  This is for
+        ///   information purposes only.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result is
+        ///   the contents of block.
+        /// </returns>
+        /// <remarks>
+        ///   Other peers are informed that the block is needed by this peer. Hopefully,
+        ///   someone will forward it to us.
+        ///   <para>
+        ///   Besides using <paramref name="cancel"/> for cancellation, the 
+        ///   <see cref="Unwant"/> method will also cancel the operation.
+        ///   </para>
+        /// </remarks>
+        public Task<IDataBlock> WantAsync(Cid id, MultiHash peer, CancellationToken cancel)
+        {
+            if (_log.IsDebugEnabled)
+            {
+                _log.Debug($"{peer} wants {id}");
+            }
+
+            var tsc = new TaskCompletionSource<IDataBlock>();
+            var want = _wants.AddOrUpdate(
+                id,
+                (key) => new WantedBlock
+                {
+                    Id = id,
+                    Consumers = new List<TaskCompletionSource<IDataBlock>> {tsc},
+                    Peers = new List<MultiHash> {peer}
+                },
+                (key, block) =>
+                {
+                    block.Peers.Add(peer);
+                    block.Consumers.Add(tsc);
+                    return block;
+                }
+            );
+
+            // If cancelled, then the block is unwanted.
+            cancel.Register(() => Unwant(id));
+
+            // If first time, tell other peers.
+            if (want.Consumers.Count != 1)
+            {
+                return tsc.Task;
+            }
+            
+            var _ = SendWantListToAllAsync(new[] {want}, full: false);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the unused local variable '_'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)